### PR TITLE
common: llama_load_model_from_url using --model-url

### LINKED ISSUE #6085: Download model gguf file from hf
### Motivation

Since GGUF is officially supported on [huggingface](https://huggingface.co/docs/hub/gguf) <3, '--model' might check for local path then fall back to hf repo id.
If a '--file' is provided it should override default for f16.

### References
- #4735 
- #5501
- https://huggingface.co/docs/hub/gguf

### PR DESCRIPTION
**Pull request by phymbert**
### Motivation

Since GGUF is officially supported on [huggingface](https://huggingface.co/docs/hub/gguf) it can be useful to start the cli directly from a http url.

We considered this at some point #4735, but it opens a lot of security concerns due to the core library having to execute commands.
So we settled with #5501 which should be good enough for most cases, but when there is no `shell` available it cannot work (example [docker distroless container](https://github.com/GoogleContainerTools/distroless)).

### Changes
- `llama_load_model_from_url` to first download the file if it does not exist locally or the remote is newer, then call `llama_load_model_from_file` in `common`
- introduce `--model-url` which will trigger the download to `--model` in `llama_init_from_gpt_params`
- enable this feature if `-DLLAMA_CURL=ON` in cmake and make toolchains
- server tests using this parameter in `embeddings.feature`

### Attention points
- The `common` will be dynamic linked to [libcurl](https://curl.se/libcurl/).
- It can later be moved to `llama` API if there is positive feedback from the community.
- it saved `${model_path}.etag` or `${model_path}.lastModified` files along with the `model_path`
- It's not possible to pass a custom CA, system CAs are always used in this first implementation on both posix and windows platform

### Task
- [x] download is triggered only if the file has changed based on `etag` and `last-modified` http headers
- [x] CI build is [passing](https://github.com/phymbert/llama.cpp/actions/runs/8317225840)
- [x] Server CI build is [passing](https://github.com/phymbert/llama.cpp/actions/runs/8317213944)

### References
- #4735
- #5501
- Closes #6085

### CODE REVIEW

**Comment by phymbert** [comment]
@ggerganov Georgi, if you approve the proposal with libcurl dependency I can continue further to support `Etag` headers to avoid downloading the file each time.

File: common/common.cpp
Comment (ggerganov) [inline common/common.cpp:19]: Change `HAVE_CURL` to `LLAMA_USE_CURL`

**Review by ggerganov — APPROVED** [review]
Yes, this can work. The `curl` dependency is optional, the implementation is isolated in `common`. Looks great

**Comment by Artefact2** [comment]
Supporting split GGUF files (for models above 50GB) would be nice too. Ideally you'd merge files with `copy_file_range()` when available to avoid extra disk I/O.

**Comment by phymbert** [comment]
> Supporting split GGUF files (for models above 50GB) would be nice too. Ideally you'd merge files with `copy_file_range()` when available to avoid extra disk I/O.

Yes I am working on this in a separate branch:
- #6000

**Comment by phymbert** [comment]
Someone can help with the error in [the windows build ](https://github.com/ggerganov/llama.cpp/actions/runs/8307457029/job/22736668256?pr=6098)?

```
D:\a\llama.cpp\llama.cpp\common\common.cpp(906,52): error C1061: compiler limit: blocks nested too deeply [D:\a\llama.cpp\llama.cpp\build\common\common.vcxproj]
```

Also @ggerganov please double check my `Makefile` changes as I am not used to.

**Comment by dranger003** [comment]
> Someone can help with the error in [the windows build ](https://github.com/ggerganov/llama.cpp/actions/runs/8307457029/job/22736668256?pr=6098)?
> 
> ```
> D:\a\llama.cpp\llama.cpp\common\common.cpp(906,52): error C1061: compiler limit: blocks nested too deeply [D:\a\llama.cpp\llama.cpp\build\common\common.vcxproj]
> ```
> 
> Also @ggerganov please double check my `Makefile` changes as I am not used to.

I submitted a PR https://github.com/ggerganov/llama.cpp/pull/6101 to fix this issue.
If you are in a hurry, you can try this workaround https://github.com/ggerganov/llama.cpp/pull/6096

File: Makefile
Comment (ggerganov) [inline Makefile:598]: The pattern is: `LLAMA_XXX=1 make` -> `-DLLAMA_USE_XXX`

```suggestion
ifdef LLAMA_CURL
```

**Comment by phymbert** [comment]
> I submitted a PR #6101 to fix this issue. If you are in a hurry, you can try this workaround #6096

Thanks a lot, no let's wait for your PR to be merged first.

File: examples/server/README.md
Comment (ggerganov) [inline examples/server/README.md:23]:
```suggestion
- `-mu MODEL_URL --model-url MODEL_URL`: Specify a remote http url to download the file (e.g https://huggingface.co/ggml-org/models/resolve/main/phi-2/ggml-model-q4_0.gguf).
```

File: common/common.cpp
Comment (ggerganov) [inline common/common.cpp:1806]:
```suggestion
struct llama_model * llama_load_model_from_url(const char * /*model_url*/, const char * /*path_model*/,
```

File: common/common.cpp
Comment (ggerganov) [inline common/common.cpp:1788]:
```suggestion
            auto * last_modified_file = fopen(last_modified_path, "w");
```

File: common/common.cpp
Comment (ggerganov) [inline common/common.cpp:1778]:
```suggestion
            auto * etag_file = fopen(etag_path, "w");
```

File: common/common.cpp
Comment (ggerganov) [inline common/common.cpp:1738]:
```suggestion
        auto * outfile = fopen(path_model, "wb");
```

File: common/common.cpp
Comment (ggerganov) [inline common/common.cpp:1714]:
```suggestion
            const char * last_modified_prefix = "last-modified: ";
```

File: common/common.cpp
Comment (ggerganov) [inline common/common.cpp:1709]:
```suggestion
            const char * etag_prefix = "etag: ";
```

File: common/common.cpp
Comment (ggerganov) [inline common/common.cpp:1706]:
```suggestion
        auto header_callback = [](char * buffer, size_t /*size*/, size_t n_items, void * userdata) -> size_t {
```

File: common/common.cpp
Comment (ggerganov) [inline common/common.cpp:1689]:
```suggestion
        auto * f_last_modified = fopen(last_modified_path, "r");
```

File: common/common.cpp
Comment (ggerganov) [inline common/common.cpp:1682]:
```suggestion
        auto * f_etag = fopen(etag_path, "r");
```

File: common/common.cpp
Comment (ggerganov) [inline common/common.cpp:1643]:
```suggestion
struct llama_model * llama_load_model_from_url(const char * model_url, const char * path_model,
```

File: examples/main/README.md
Comment (ggerganov) [inline examples/main/README.md:70]:
```suggestion
-   `-mu MODEL_URL --model-url MODEL_URL`: Specify a remote http url to download the file (e.g https://huggingface.co/ggml-org/models/resolve/main/phi-2/ggml-model-q4_0.gguf).
```

**Review by ggerganov — APPROVED** [review]

File: common/common.cpp
Comment (ngxson) [inline common/common.cpp:1678]: Maybe we can use snprintf here

File: common/CMakeLists.txt
Comment (ngxson) [inline common/CMakeLists.txt:51]: Please correct if I’m wrong, but here we check if curl is found on build system, then we turn on LLAMA_CURL. I wonder in case user have libcurl but don’t want to enable this feature in llama.cpp (to make sure there’s no depenency for example), then can we do that ?

Comment (phymbert) [reply common/CMakeLists.txt:51]: AFAIK It's dynamic linking so it will trigger linking to `libcurl` at runtime only if user use `--model-url` or `llama_load_model_from_url` and at this stage I think it's OK. If users complain that  `ldd` is listing `libcurl.so` we can later on introduce `LLAMA_CURL=ON` cmake option as in the make build toolchain.

Comment (ngxson) [reply common/CMakeLists.txt:51]: At least on Linux that’s not the case. If the libcurl.so is listed in ldd, than it will be dynamically linked the moment we run the binary (even if the function is not used). In fact, for this reason, I avoid having dynamic link when possible (sometimes it breaks the portablity due to mismatch version of library in build time & runtime). Also I think in Makefile we only build with libcurl only if specified, so should we match the behavior of cmake & makefile?

**Review by ngxson — APPROVED** [review]
Looks good to me, thanks for having implementing this!

One point not important, maybe done in the future: it would be nice to have some kind of progress bar and to show current download speed.

Also regarding #6000, if we decide to store number of splits in a metadata, that may break this feature because we don't know in advance how many parts of the model to be download. To fix that, maybe we can have first part of model only contains metadata (no tensor) - will weight only few MBs, then we extract metadata to see how many parts in total need to be downloaded.

File: common/common.cpp
Comment (slaren) [inline common/common.cpp:60]: `windows.h` is already included before.

**Comment by phymbert** [comment]
Need to fix the windows server CI tests, it's not passing at the moment

**Comment by phymbert** [comment]
It finally works on both windows and linux CI tests. Happy to merge it when CI passes

**Comment by phymbert** [comment]
@ggerganov Georgi, I suffered a little on Windows, but this implementation works on both platform. Would you please do another review as I changed a little bit the logic - and sorry but still need to find a good linter for CLion.

**Review by ggerganov — APPROVED** [review]

File: examples/server/tests/features/environment.py
Comment (cebtenzzre) [inline examples/server/tests/features/environment.py:83]: @phymbert This calls `TerminateProcess(handle, SIGTERM)` on Windows and `os.kill(pid, SIGKILL)` on Unix. `os.kill(pid, signal.SIGTERM)` also calls TerminateProcess on Windows. psutil really seems like overkill here.

Comment (phymbert) [reply examples/server/tests/features/environment.py:83]: Again,feel free to issue a PR

**Pull request merged**
